### PR TITLE
fix: #39 - Improve constructor performance

### PR DESCRIPTION
**Pull request by realharry**
REF: https://github.com/ipld/js-cid/issues/39

Updated the isCID() implementation to use a (private) intermediate method to validate CID without relying on the error being thrown during validation.

The validateCID() method also uses this intermediate method, and its behavior has been preserved (e.g., with regards to throwing an error for invalid CIDs).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "cids",
-  "version": "0.5.2",
+  "version": "0.5.3",
```
Comment (diasdavid) [inline package.json]: @realharry version gets updated on publish. Please revert this change.

File: package.json [hunk 2]
```diff
@@ -43,6 +43,10 @@
     "aegir": "^12.0.8",
     "chai": "^4.1.2",
     "dirty-chai": "^2.0.1",
+    "karma-mocha": "^1.3.0",
+    "karma-mocha-webworker": "^1.3.0",
```
Comment (diasdavid) [inline package.json]: This gets installed with `aegir`. Did it not work for you?

File: package.json [hunk 3]
```diff
@@ -43,6 +43,10 @@
...
+    "mocha": "^5.0.1",
```
Comment (diasdavid) [inline package.json]: Same as above

File: package.json [hunk 4]
```diff
@@ -43,6 +43,10 @@
...
+    "multihashing": "^0.3.2",
```
Comment (diasdavid) [inline package.json]: We only use `multihashing-async` to use WebCrypto in the browser.

File: src/index.js [hunk 5]
```diff
@@ -219,12 +219,11 @@ class CID {
    */
   static isCID (other) {
     try {
```
Comment (dignifiedquire) [inline src/index.js]: why is this `try` block still here? shouldn't this be just `return Boolen(CID._checkCIDComponents(other))` now?

File: package.json (hunk 4 above)
Comment (realharry) [reply package.json]: I was "experimenting", and likely I forgot to remove this package after (accidentally) adding it. I'll remove it.

File: package.json (hunk 2 above)
Comment (realharry) [reply package.json]: @diasdavid It was working on my machine, but the CI builds were failing (especially, Jenkins) as shown in my checkin history. I added these to make the CI builds work. I'll revert it back.

File: src/index.js (hunk 5 above)
Comment (realharry) [reply src/index.js]: @dignifiedquire If you look at the original implementation of `validateCID()`, it relies on the behavior of `mh.validate(other.multihash)`, which is supposed to throw an error for invalid multihash. That behavior is preserved in the updated `validateCID()`. And, in order to be able to use/share a common implementation between `isCID()` and `validateCID()`, I had to preserve that behavior in `_checkCIDComponents()` as well. Therefore `isCID()` still needs to catch the error to return false/true _in case_ an error is thrown (e.g., from `mh.validate(other.multihash)`). Contractually (in terms of interface), `isCID()` never throws an error since we now have an "outer" `try/catch` containing its implementation (even through future updates). I hope my answer makes sense.

Comment (dignifiedquire) [reply src/index.js]: Ah I forgot that. This is unfortunate, `mh.validate` should at least have a counter part that doesn't throw.

File: package.json (hunk 4 above)
Comment (realharry) [reply package.json]: Oh, I must have misunderstood your comment. I'll remove `multihashing-async`, but I will still need `multihashing` in dev-dependencies. I use `multihashing` in my test case `cidperf-a.js`.

**Comment by realharry** [comment]
I removed the karma/mocha dev-dependencies. Getting the same build error as before. @diasdavid Any suggestions?

File: src/index.js (hunk 5 above)
Comment (realharry) [reply src/index.js]: Yes. but, as I indicated in my comment on #39, the try/catch itself is not an issue. It's the cost of creating an error object in the heap that takes time. Depending on the use cases, _if_ we rarely hit `mh.validate` part in the "if chaining", the cost of having that is not much. The question is, out of 100 times or a million times calling `isCID()` in a typical use case (e.g., are CIDs generally expected to be valid in most cases?, etc.), how many times we end up creating the error object? Having said that, I do agree with you. The code will be cleaner if we don't have to include `try/catch`. Thanks.

**Comment by realharry** [comment]
Sorry for the mixup. All CI builds pass at this point. Thanks.

File: src/index.js [hunk 6]
```diff
@@ -235,23 +234,39 @@ class CID {
...
     }
 
     if (typeof other.codec !== 'string') {
-      throw new Error('codec must be string')
+      return 'codec must be string'
     }
 
     if (!Buffer.isBuffer(other.multihash)) {
-      throw new Error('multihash must be a Buffer')
+      return 'multihash must be a Buffer'
```
Comment (diasdavid) [inline src/index.js]: These should be thrown errors.

File: src/index.js [hunk 7]
```diff
@@ -235,23 +234,39 @@ class CID {
...
+
+  /**
+   * Test if the given input is a valid CID object.
+   * Returns an error message if it is not, or
+   * Throws an error its multihash is invalid.
+   *
+   * @param {any} other
+   * @returns {string}
+   */
+  static _checkCIDComponents (other) {
```
Comment (diasdavid) [inline src/index.js]: This function can be extracted to a separate file given it is an utility.

File: test/profiling/alt-impl/cid1.js [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+      throw new Error('multihash must be a Buffer')
+    }
+
+    mh.validate(other.multihash)
+  }
+}
+
+CID1.codecs = codecs
+
+module.exports = CID1
```
Comment (diasdavid) [inline test/profiling/alt-impl/cid1.js]: Can we call it CID_A and CID_B given that we already use 0 and 1 for version numbers? Pretty cool to include profiling tests!

Comment (diasdavid) [reply test/profiling/alt-impl/cid1.js]: Thinking more about this. Instead of duplicating the code. This could just be installing the CID version on npm.

File: test/profiling/alt-impl/cid2.js [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+    mh.validate(other.multihash)
+    return null
+  }
+}
+
+CID2.codecs = codecs
+
+module.exports = CID2
```
Comment (diasdavid) [inline test/profiling/alt-impl/cid2.js]: Why duplicate instead of using the version in src?

Comment (realharry) [reply test/profiling/alt-impl/cid2.js]: @diasdavid Thanks for pointing that out. I wasn't really happy with this solution, but I couldn't come up with a better one. The issue was: (1) I wanted to use, _as a reference,_ a "snapshot" of the CID implementation as of, say, Sunday. The CID implementation will (likely) change over time, and using the "current" version (in src) will not make sense for this particular test (say, 2 months from now). (2) I wanted to try a different "alternative" implementations _for comparison._ And, putting alternative (trial) implementations in the prod source code (e.g., as an extra method, etc.) would not have made sense. 

That's why I ended up duplicating the code for _both._ Obviously, there is something "funny" about doing it this way. I'm not sure what would be the best way though. Any suggestions?

File: test/profiling/alt-impl/cid1.js (hunk 8 above)
Comment (realharry) [reply test/profiling/alt-impl/cid1.js]: Doing it this way will require creating a new npm project folder (with package.json) for testing/profiling. This only works for the "reference" impl. We may still need a separate/duplicated CID for comparison. (See the comment below.)

File: src/index.js (hunk 7 above)
Comment (realharry) [reply src/index.js]: I can create a file `cid-util.js` or `util.js` or `util/cid-util.js`. Any preferences?

File: src/index.js (hunk 6 above)
Comment (realharry) [reply src/index.js]: OK.

File: test/profiling/alt-impl/cid2.js (hunk 9 above)
Comment (realharry) [reply test/profiling/alt-impl/cid2.js]: In hindsight, I think your suggestion, if I understand correctly, seems to make more sense. I now have to reverse reference and alternative (in my mind):

* Reference: CID version X (in NPM)
* Alternative: CID in src (with alternative/trial implementation)

(When I started out, the CID in src was the reference, and "my version" was the alternative.)

Comment (diasdavid) [reply test/profiling/alt-impl/cid2.js]: You can npm install a locked version on the profiling test folder (therefore one version) and require the other from the src folder (and you get the other version to compare). Makes sense?

**Comment by realharry** [comment]
Updated per feedback:
(1) Created a new cid-util.js and its corresponding test spec.
(2) Simplified the test runner so that it now takes only one CID implementation. Will need to run the test multiple times for each alternative implementation.

File: src/cid-util.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+   * Test if the given input is a valid CID object.
+   * Returns an error message if it is not, or
+   * Throws an error its multihash is invalid.
+   *
+   * @param {any} other
+   * @returns {string}
+   */
+  checkCIDComponents: function (other) {
+    if (other == null) {
+      return 'null values are not valid CIDs'
```
Comment (diasdavid) [inline src/cid-util.js]: This should also be a throw Error

File: src/cid-util.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    if (!(other.version === 0 || other.version === 1)) {
+      return 'Invalid version, must be a number equal to 1 or 0'
```
Comment (diasdavid) [inline src/cid-util.js]: plus this one

File: src/cid-util.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    if (typeof other.codec !== 'string') {
+      return 'codec must be string'
```
Comment (diasdavid) [inline src/cid-util.js]: plus this one

**Comment by realharry** [comment]
@diasdavid Jenkins build fails on Mac. Is there anything I can do to fix this?

File: src/cid-util.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    if (!Buffer.isBuffer(other.multihash)) {
+      throw new Error('multihash must be a Buffer')
```
Comment (diasdavid) [inline src/cid-util.js]: Just like this one :)

File: src/cid-util.js [hunk 14]
```diff
@@ -0,0 +1,36 @@
+'use strict'
+
+const mh = require('multihashes')
+
+var CIDUtil = {
+  /**
+   * Test if the given input is a valid CID object.
+   * Returns an error message if it is not, or
+   * Throws an error its multihash is invalid.
```
Comment (diasdavid) [inline src/cid-util.js]: Let's simplify it and just make it throw an error if it is doesn't have valid CID components, the error message will tell which component is invalid.

Comment (realharry) [reply src/cid-util.js]: @diasdavid I'm a little bit confused. 

The new implementation of `validateCID()` already does that. And, this behavior has been preserved from the existing implementation. The purpose of `checkCIDComponents()` was to use the same logic in `isCID()` without relying on the error being thrown. The idea was that `isCID()` was slow because of the error objects being created on the heap (which obviously we need to verify in the long run). If we change `checkCIDComponents()` to throw errors for all these conditions, we are effectively going back to the old implementation where `isCID()` directly uses `validateCID()`. The purpose of `checkCIDComponents()` was to avoid using error objects in CID validation in the constructor (via `isCID()`).

Please advise.

Comment (diasdavid) [reply src/cid-util.js]: Oh, I see.

If that is the case. Let's wrap the mh.validate on a tryCatch block inside the checkCIDComponents itself and only make that function return strings or undefined (should enable V8 to optimize it even more).

Comment (realharry) [reply src/cid-util.js]: OK. I'll check in the updated changes shortly.

File: src/cid-util.js (hunk 11 above)
Comment (diasdavid) [reply src/cid-util.js]: same as https://github.com/ipld/js-cid/pull/45/files#r173727533

File: src/cid-util.js (hunk 12 above)
Comment (diasdavid) [reply src/cid-util.js]: same as https://github.com/ipld/js-cid/pull/45/files#r173727533
